pkg/controller/pxc/upgrade.go

```diff
@@ -119,106 +66,82 @@ func (r *ReconcilePerconaXtraDBCluster) updatePod(ctx context.Context, sfs api.S
-		logCollectorC, err := sfs.LogCollectorContainer(cr.Spec.LogCollector, cr.Spec.LogCollectorSecretName, secretsName, cr)
+	err = k8sretry.RetryOnConflict(k8sretry.DefaultRetry, func() error {
+		currentSet := sfs.StatefulSet()
+		err := r.client.Get(ctx, types.NamespacedName{Name: currentSet.Name, Namespace: currentSet.Namespace}, currentSet)
 		if err != nil {
-			return errors.Wrap(err, "logcollector container error")
+			return errors.Wrap(err, "failed to get statefulset")
 		}
-		if logCollectorC != nil {
-			newContainers = append(newContainers, logCollectorC...)
-		}
-	}
+		annotations := currentSet.Spec.Template.Annotations
+		labels := currentSet.Spec.Template.Labels
 
-	// volumes
-	sfsVolume, err := sfs.Volumes(podSpec, cr, r.getConfigVolume)
-	if err != nil {
-		return errors.Wrap(err, "volumes error")
-	}
+		sts, err := pxc.StatefulSet(ctx, r.client, sfs, podSpec, cr, secrets, initImageName, r.getConfigVolume)
+		if err != nil {
+			return errors.Wrap(err, "construct statefulset")
+		}
 
-	// application container
-	appC, err := sfs.AppContainer(podSpec, secretsName, cr, sfsVolume.Volumes)
-	if err != nil {
-		return errors.Wrap(err, "app container error")
-	}
+		// support annotation adjustements
+		pxc.MergeMaps(annotations, sts.Spec.Template.Annotations, newAnnotations)
 
-	newContainers = append(newContainers, appC)
+		pxc.MergeMaps(labels, sts.Spec.Template.Labels)
 
-	if len(initContainers) > 0 {
-		newInitContainers = append(newInitContainers, initContainers...)
-	}
+		annotations["percona.com/configuration-hash"] = configHash
 
-	if podSpec.ForceUnsafeBootstrap {
-		log.Info("spec.pxc.forceUnsafeBootstrap option is not supported since v1.10")
+		// change TLS secret configuration
+		if sslHash != "" && cr.CompareVersionWith("1.1.0") >= 0 {
+			annotations["percona.com/ssl-hash"] = sslHash
+		}
+		if sslInternalHash != "" && cr.CompareVersionWith("1.1.0") >= 0 {
+			annotations["percona.com/ssl-internal-hash"] = sslInternalHash
+		}
 
-		if cr.CompareVersionWith("1.10.0") < 0 {
-			ic := appC.DeepCopy()
-			ic.Name = ic.Name + "-init-unsafe"
-			ic.Resources = podSpec.Resources
-			ic.ReadinessProbe = nil
-			ic.LivenessProbe = nil
-			ic.Command = []string{"/var/lib/mysql/unsafe-bootstrap.sh"}
-			newInitContainers = append(newInitContainers, *ic)
+		if vaultConfigHash != "" && cr.CompareVersionWith("1.6.0") >= 0 && !isHAproxy(sfs) {
+			annotations["percona.com/vault-config-hash"] = vaultConfigHash
 		}
-	}
 
-	// sidecars
-	sideC, err := sfs.SidecarContainers(podSpec, secretsName, cr)
-	if err != nil {
-		return errors.Wrap(err, "sidecar container error")
-	}
-	newContainers = append(newContainers, sideC...)
+		if envVarsHash != "" {
+			annotations["percona.com/env-secret-config-hash"] = envVarsHash
+		}
 
-	newContainers = api.AddSidecarContainers(log, newContainers, podSpec.Sidecars)
+		if isHAproxy(sfs) && cr.CompareVersionWith("1.6.0") >= 0 {
```

Comment:
we can remove this version check too

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/1743/commits/f8545394952bdd75904a101ebb697d6755cfbd81